### PR TITLE
[FOR PROGRESS, DO NOT MERGE] Rewrite FirebaseUI auth

### PR DESCRIPTION
**Pull request by SUPERCILEX**
#856, #948, #526, future support for #724 and #661, easier #123 implementation. In addition, every config change bug you can think of that people haven't reported.

cc: @samtstern + another human?

I was going to name this PR something more specific like "Fix all config change bugs," but then I saw the LoC changed... sorry @samtstern! 😊 

## Intro

Basically, if you try rotating your screen on pre-PR FirebaseUI, either the user is half signed-in but the UI doesn't show it, or we crash, or we just break randomly. To be frank, we _suck_! For example, rotate the screen after clicking the email sign up button, the user will actually be signed-in, but we're now broken. Click the button again, and we'll say the user already has an existing account. Duh, they just signed up! 😄

If you can't already tell, I'm super excited about this one which is why I'm sharing it early. So far, it just works! Plain and simple, it _works_. On the emulator, I've been mashing the Ctrl + Left/Right buttons to create config change after config change and we nail the sign in every single time. Every. Single. Time!

Anyway, the post-PR world will be a new era of ultimate stability. 😄

TODO:
- [x] An event flow for progress changes and a `ProgressBar` under the `Toolbar`
- [x] Rewrite tests (shouldn't be testing UI, only business logic)
- [ ] Add an `AuthExcecutor` and run all `Task`s synchronously
- [ ] https://firebase.google.com/docs/reference/android/com/google/firebase/auth/FirebaseAuthUserCollisionException
- [x] Refactor phone auth, it's completely broken
- [x] Move `SignInHandler` to a service so we can get rid of the static fields? Nope. I'm open to reconsideration, but this seems like a lot of extra work for little benefit.

## The Problem

Fundamentally, most architectures assume that data doesn't talk back to UI. If we think of the auth process of signing-in and saving credentials etc. as "data," then we run into a problem: the data now needs to tell the UI to do stuff. This means both the user and the data are going to be fighting for control over the UI and need to somehow stay connected. In pre-PR FirebaseUI, this lead to nightmares where we just kinda gave up and ignored auth results if the user rotated their screen.

## The Solution

I've used a mix of an eventbus + Rx model. Basically, all background auth stuff exists independently from the UI. However, since we need to talk to the UI, we use 4 data flow tracks: one for starting an `Intent`, another for starting a `PendingIntent`, yet another one for sending `IdpResponse`s back to the UI to decide what to do, and the last one to send activity results. No more forwarding `onActivityResult` everywhere! 😄 

This means `setRetainInstance(true)` fragments are dead, the memory leak hacks in `FacebookProvider` and whatnot are dead, the `SaveSmartLock` fragment is dead, etc.

I was hoping to be able to submit my PR in more bite friendly chunks, but everything was such a mess that it just wasn't possible.

### Project structure overview

- `AuthViewModelBase` and its descendents have access to `FirebaseAuth` (and only them)
  - `SignInHandler` manages _all_ sign-in business login from failures to saving credentials to merging accounts etc.
  - `SignInKickstarter` manages the intro stuff like figuring out where to send users and the opening credential picker
- `Provider`s are now UI only and have their companion `ProviderHandler`s. For example `GoogleSignInHandler` manages all the logic for the `GoogleApiClient` and the Facebook and Twitter ones manage retrieving profile info, etc.

The rest is essentially the same, just forwarding everything to those classes above. I did rewrite the `GoogleApiClient` managers and error handling.

### Potential limitations

The only one I can think of is perhaps callback hell. I'm not sure how to fix that and for now, I just want it to work. However, I really want it done _right_. So @samtstern and anyone else, if there's anything you think we can do better or another better solution you can think of, please tell me! I want auth to be a solved problem once and for all! 😄

@samtstern Since this is a massive PR, I'm expecting review to possibly take up to a month, so just take your time with it. It's got to be done right. I'll update it with a review of the classes and changes once I finish the TODOs listed above, but for now you can familiarize yourself with it. Enjoy! 😄

### CODE REVIEW

**Comment by samtstern** [comment]
@SUPERCILEX before I continue I just want to say it's awesome that you're taking on such huge task!  I agree with your diagnosis of the problem(s).

Before we go any further, we need to consider how to actually get this done.  One giant PR is just not a reasonable way to refactor the whole module.  We just barely managed that for "the big Kahuna" database PR and that became quite a mess at the end.  I think we need to:

- Break this down into smaller tasks.
- File an issue for each one, with justification and some indication of what the fix will look like.
- Write/review the PRs one at a time.

If this was an internal project at Google I'd say we need a design doc and a plan of action.  If you want, we could collaborate in a Google doc or something that supports better back-and-forth iteration than Github comments.

We should also discuss any breaking API changes that this will involve separately.  I'm open to those, but we should consider them separately from the implementation.

Apologies to bring in all this bureaucracy but I think it's the only effective way to achieve something this large.

**Comment by SUPERCILEX** [comment]
@samtstern Yeah, I totally understand. AFAIK, there are no breaking changes. I like the Google doc idea, but I don't have time to continue working on this PR right this sec. Sometime by the end of the week I'll get those TODOs done and start working on a Google doc.

TBH, I'm really not sure how we can break this up since everything was so connected. I'll start thinking about it, but maybe we can setup a live chat or video conference to talk about it or something? Maybe it'll all work out and I'll think of something, but for now I'm stumped as to how we can break this apart while keeping working code between the intermediary steps. Also, the tests are going to be screwed. 😆

Anyway, I'll get something going and get back to you sometime by the end of the week.

Thanks for finding a plan of action! 😄

**Comment by samtstern** [comment]
@SUPERCILEX happy to do a video chat if you want.  I think a basic design doc of some form is the first step to an attack plan.

I'm actually fine if we have broken code on a branch at some points, at least it will be broken intentionally!

**Comment by SUPERCILEX** [comment]
@samtstern SGTM!

**Comment by SUPERCILEX** [comment]
Phone auth refactor is done! I've got to say, props to people who do this on their own—it's hard as a brick! 😆

Also, I've added support for phone numbers in Smart Lock which will need to be discussed (I did it to see if it was possible, not sure if it's a desired change though).

**Comment by SUPERCILEX** [comment]
@samtstern Ok, I've got a [design doc going](https://docs.google.com/document/d/1fI1PdF_dkf4ij3QAOiGzIS2FksNDdN7sZ53CZPXjgPc/edit?usp=sharing) which I've given you editing access and comment access for everyone else. For the 235 people watching this repo, if anyone is interesting in architecture, loading states, or testing, please have a look at the doc and give me feedback! 😄 

@samtstern If there's anything else you'd like me to include in the doc, please let me know. Otherwise, I'm thinking of a PR flow like this:
1. The GoogleApiClient manager rewrites
1. `SignInHandler` and its related flow observables
1. The picker activity with the new provider handler stuff
1. The intro flow
1. Welcome back stuff
1. And the Big Kahuna, phone number auth

That flow would let us build up dependencies though we'd still get a bunch of compile errors in the meantime. What do you think?

**Comment by TheCraftKid** [comment]
Wow, how didn't I notice this?
This is going to lead to a way better than the solution I had whipped up to fix #806. Going through the old auth code sort of gave me chills. Only @SUPERCILEX would try to rewrite the auth module. :smile: 
My thoughts about the changes are in the Google Doc.

**Comment by samtstern** [comment]
@SUPERCILEX / @TheCraftKid  ok I am starting to wrap my head around some of this!  I think the general ideas are good (or at least the parts I can understand). 

Procedurally, here are a few things to know about:

   * We've got some big partners moving to FirebaseUI (from Digits) by the end of the month.  So even if this was all 100% gold we can't make any big changes until at least early November and if those big partners find show-stopper bugs I will have to prioritize fixing them on top of the existing codebase.
   * I will personally be doing a mix of work travel and personal travel from Oct 27th --> November 12th.  I'll still be on Github a lot of that time but at odd hours.
   * There are some changes coming to the Google Sign In API soon that I think will make you happy and also remove the need for the `GoogleApiClient` manager classes.  If you want to know more we can talk.

Beyond that let's do most discussion in the doc!

**Comment by SUPERCILEX** [comment]
@samtstern Ok, that's too bad.

Once we figure out a PR strategy, could I still open them and you just wait until you're ready to start merging them again? BTW, it's really cool see that partners are moving to FirebaseUI—it makes our work here even more valuable! Too bad there are going to be a bunch of config change bugs though. 😢😆

Thanks for not leaving me in the dark regarding your GitHub availability.

Any chance I could get alpha access to the new sign-in APIs? They sound like exciting! (I'll take a wild guess the `Task` API  will be involved... whaaaaaat? Google using its own `Task` API, who would have thought!? 😆)

@samtstern cool beans, thanks for all the info!

**Comment by samtstern** [comment]
@SUPERCILEX yes we can definitely open PRs at any time, I just didn't want you to think I was ignoring them.

**Comment by SUPERCILEX** [comment]
@samtstern sweet, SGTM!

File: constants.gradle [hunk 1]
```diff
@@ -12,5 +12,5 @@ project.ext {
 
     firebaseVersion = '11.4.2'
     supportLibraryVersion = '26.1.0'
-    architectureVersion = '1.0.0-beta1'
+    architectureVersion = '1.0.0-beta2'
```
Comment (TheCraftKid) [inline constants.gradle]: Bump to 1.0.0-rc1

File: auth/src/main/java/com/firebase/ui/auth/data/remote/ProfileMerger.java [hunk 2]
```diff
@@ -39,7 +36,7 @@ public ProfileMerger(IdpResponse response) {
             return Tasks.forResult(authResult);
         }
 
-        User user = mIdpResponse.getUser();
+        User user = mResponse.getUser();
         if (TextUtils.isEmpty(name)) { name = user.getName(); }
```
Comment (TheCraftKid) [inline auth/src/main/java/com/firebase/ui/auth/data/remote/ProfileMerger.java]: Nit: put contents of if on separate lines

**Comment by samtstern** [comment]
@SUPERCILEX while I am also hopeful that this will solve all of our config bugs let's not promise the moon to those filing issues like #856, we have no idea when this refactor will land and how well it will work until we're done :-)

**Comment by SUPERCILEX** [comment]
@samtstern True, true. My B. 😊 I guess I'm just super hopeful since Architecture Components guarantees that no events are emitted after `onStop`. 🤞 😄

**Comment by samtstern** [comment]
@SUPERCILEX I haven't had a chance to look at your most recent changes to the Google Doc since I've been out of the office but check out the new changes to Google Sign In and SmartLock in `11.6.0`.  I think these two changes should allow us to stop using `GoogleApiClient` which will be a *HUGE* step to a rotation-bug-free world.

**Comment by samtstern** [comment]
I am happy to lead the way on the GoogleApiClient removal, but want to make sure you're aware and not get in the way if you're already thinking about this change.

**Comment by SUPERCILEX** [comment]
@samtstern Holy crap, oh boy oh boy!!! 😄 I think this merited a major version bump to `v12.x`, but anyway... how soon do you want to kill the `GoogleApiClient`s? This PR goes a long way already: there's basically only the [`GoogleApiConnector`](https://github.com/SUPERCILEX/FirebaseUI-Android/blob/247b69feb68bfdf1bb45c6b61b5d4c50549e8099/auth/src/main/java/com/firebase/ui/auth/util/data/remote/GoogleApiConnector.java) and its subclasses class to kill. For opening those credential dialogs, I had already refactored them [not to use connections](https://github.com/SUPERCILEX/FirebaseUI-Android/blob/247b69feb68bfdf1bb45c6b61b5d4c50549e8099/auth/src/main/java/com/firebase/ui/auth/ui/phone/CheckPhoneNumberHandler.java#L75-L79). And for saving credentials, we'd just have to inject that [here](https://github.com/SUPERCILEX/FirebaseUI-Android/blob/247b69feb68bfdf1bb45c6b61b5d4c50549e8099/auth/src/main/java/com/firebase/ui/auth/data/remote/SignInHandler.java#L89-L91).

If getting rid of those pesky `GoogleApiClient`s is the priority, I can deal with the merge conflicts. Otherwise, we could lop that into this PR since it already gets rid of most of our existing connection infrastructure.

What do you think?

**Comment by samtstern** [comment]
@SUPERCILEX I wouldn't say getting rid of GAC is the priority, but it seems like a good self-contained piece of work that will help in making this PR smaller.  My main goal is to find ways to make this refactor a series of medium-sized improvements rather than another Big Kahuna!

Getting rid of GAC would:
  * Make all of our SmartLock logic wayyy simpler
  * Reduce some fragment use
  * Move everything to `Task` which would help with our chaining situation

**Comment by SUPERCILEX** [comment]
@samtstern Yep, that's exactly what this PR does. 😄 There are no `setRetainInstance(true)` fragments and we only have a single point of entry for all of our sign in code—so adding the smart lock `save(...)` to the chain would take just a few lines of code in one place vs the everywhere if we tried to do it today.

I totally feel you about making this PR smaller and I just realized I forgot to share my thoughts on how we can accomplish this. I was thinking of cutting the PR into logical chunks while ignoring compilation errors (since you said that was ok). Once I get those TODO items done and a green build, we can keep this PR open as the ultimate source of truth and watch the lines of code changed numbers dwindle down to eventually 0. As for the order of PRs, here's what I was thinking:
1. Start with the public API i.e. the credential/Smart Lock base stuff (which is used in `AuthUI`) but not the the saving or other internal stuff. With Play Services 11.4, that PR should be must smaller and simpler.
1. Setup the basic infrastructure needed for this PR: the `LiveData`s, `HelperActivityBase` changes, etc.
1. Rewrite the `Provider`s
1. Continue with the `KickOffActivity` and boot up credential stuff. Again, this should be much simpler now with the new Play Services
1. Rewrite the `AuthMethodPickerActivity` (should be a very small PR)
1. Rewrite the email Activities/Fragments
1. Rewrite the welcome back stuff
1. Rewrite phone auth (because the phone auth code is so 😭, I don't see how to make that PR smaller and it will be like ~3,000 lines of code changed
1. Finalize any last tidbits, get everything compiling, and a green build

With the above sequence, we can build out a logical dependency tree. In addition, most PRs should be under 1,000 LoC changed since `10,000 / 9 ~= 1,000`.

BTW, all of the above should be done on a separate dev branch since it won't be compiling for a while and we want to be able to do other stuff in the meantime. Once it's got a green build, we would merge it back into the normal dev branch.

@samtstern What do you think?

**Comment by samtstern** [comment]
Sorry I think I was not clear before:

> I was thinking of cutting the PR into logical chunks while ignoring compilation errors (since you said that was ok).

I meant that it's ok if the build is red (like breaking existing tests or other strange things) but I always want the code to compile!  If we can't compile it's really hard to develop or do sanity checks along the way.  I'm sure you agree.

I'd like to see this broken up into smaller chunks each one making small improvements to the library toward the larger goal.  We could even release along the way with some of the improvements in place as long as we never go backwards.

We should really do this the right way, filing issues for changes we want to see and then sending PRs for each one after brief discussion.  This will also allow others besides you (like me, jstimes, and maybe TheCraftKid) to contribute.  The more people that contribute to this refactoring the more competent maintainers we have going forward and the better it will be.

You've done an amazing job outlining the vision and the amount of work in this PR is staggering, but we need to get some of that vision out of your head in reasonable chunks or we'll be trapped strategizing forever!

**Comment by SUPERCILEX** [comment]
@samtstern Ohhhhhhhh, I see what you're looking for!!! 😄 Man, this is going to be super painful... we'll have to figure out how to mix old and new code. Anyway, SGTM! I'll start thinking about how to slice and dice the thing and hopefully submit an initial PR sometime on Monday or earlier. 😁

**Comment by samtstern** [comment]
@SUPERCILEX awesome, yeah it may be painful but at least we will be able to wrap our heads around it and ship code!

**Comment by samtstern** [comment]
Oh and as mentioned before, I think nuking as much `GoogleApiClient` use would be a good step since a lot of our complexity comes from how we use it, could give us breathing room to get into the architecture changes that are the heart of this.

**Comment by SUPERCILEX** [comment]
@samtstern Oh boy, we're so close! I saw #1007 which is almost all I have left to do for this PR so it can become our source of truth. We should be able to rock and roll tomorrow with our first PR! 🎸😄

**Comment by SUPERCILEX** [comment]
Aw the tests failed. Sigh, they passed on my machine.

**Comment by SUPERCILEX** [comment]
@samtstern Alrighty, it's about to rain down PRs! 😄 The new GoogleApi cut chopped down that code by [half](https://github.com/firebase/FirebaseUI-Android/pull/962/commits/df87350fe8445a27820c8a8cb1c3853c5f0b5644) which is prety dang cool.

_But_ (and this is a big BUT), there's [this adorable ugliness](https://github.com/SUPERCILEX/FirebaseUI-Android/blob/df87350fe8445a27820c8a8cb1c3853c5f0b5644/auth/src/main/java/com/firebase/ui/auth/data/remote/SignInHandler.java#L218-L236). I'll explain more when we get there, but could you ping the PS team (or do it yourself since you're a Googler 😄) to add a `build()` method to the `Auth.AuthCredentialsOptions.Builder` and add an overload to the `Credentials.getClient(...)` methods that accepts an `Auth.AuthCredentialsOptions`? Mini #rant: this is why we need PS to be open-source! I know it's never going to happen for Android, but these kinds of simple oversights could be circumvented if the full breadth of the Android community were allowed to participate in PS developpement.

🚀 PRs incoming!

**Comment by samtstern** [comment]
@SUPERCILEX that reflection hack is nasty!  Raised a bug internally with the SmartLock folks.

**Comment by SUPERCILEX** [comment]
Awesome, thanks!

**Comment by SUPERCILEX** [comment]
@samtstern I didn't quite have the time to finish, but my fork is almost compiling again. 😭😆 Then I'll get a PR going that rewrites all the provider implementations hopefully around Wednesday. (spoiler alert: the gross `FacebookProvider` manual memory management is gone! 😁)

**Comment by SUPERCILEX** [comment]
All the changes left here have been moved to #1253.

**Pull request closed without merging**